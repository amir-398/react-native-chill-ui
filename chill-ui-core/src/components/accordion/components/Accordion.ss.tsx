import type { AccordionPropsSs } from '@types';

import { BoxSs } from '@components/box';
import { PropsWithChildren } from 'react';

import styles from '../styles/Accordion.ss.styles';
import { AccordionProvider } from './AccordionContext';
import { accordionDefaultProps } from '../utils/defaultProps';

/**
 * Accordion component provides a collapsible content area with support for single or multiple items.
 *
 * @example
 * ```tsx
 * <Accordion type="single" collapsible>
 *   <AccordionItem value="item-1">
 *     <AccordionTrigger>Is it accessible?</AccordionTrigger>
 *     <AccordionContent>Yes. It adheres to the WAI-ARIA design pattern.</AccordionContent>
 *   </AccordionItem>
 * </Accordion>
 * ```
 *
 * @param children - Child AccordionItem components
 * @param type - Whether single or multiple items can be open at once ('single' | 'multiple')
 * @param collapseIcon - Icon to show when item is expanded (default: 'angle-down-solid')
 * @param collapsible - Whether items can be collapsed when type is 'single' (default: false)
 * @param defaultValue - Default open items (string for single, array for multiple)
 * @param disabled - Disable the entire accordion (default: false)
 * @param expandIcon - Icon to show when item is collapsed (default: 'angle-down-solid')
 * @param hasCollapseIcon - Whether to show collapse/expand icons (default: true)
 * @param iconPosition - Position of the expand/collapse icon (default: 'right')
 * @param onValueChange - Callback fired when accordion state changes
 * @param props - Additional ViewProps for the container
 *
 * @see {@link https://github.com/your-repo/chill-ui/tree/main/src/components/accordion/README.md Documentation}
 */
export default function Accordion(props: PropsWithChildren<AccordionPropsSs>) {
  const {
    children,
    collapseIcon = accordionDefaultProps.collapseIcon,
    collapsible = accordionDefaultProps.collapsible,
    defaultValue,
    disabled = accordionDefaultProps.disabled,
    expandIcon = accordionDefaultProps.expandIcon,
    hasCollapseIcon = accordionDefaultProps.hasCollapseIcon,
    iconPosition = accordionDefaultProps.iconPosition,
    onValueChange,
    style,
    type = accordionDefaultProps.type,
  } = props;

  return (
    <AccordionProvider
      type={type!}
      collapsible={collapsible}
      defaultValue={defaultValue}
      disabled={disabled}
      hasAnimation={false}
      animationDuration={0}
      hasCollapseIcon={hasCollapseIcon ?? true}
      iconPosition={iconPosition}
      expandIcon={expandIcon}
      collapseIcon={collapseIcon}
      onValueChange={onValueChange}
    >
      <BoxSs style={[styles.accordion, style]}>{children}</BoxSs>
    </AccordionProvider>
  );
}
