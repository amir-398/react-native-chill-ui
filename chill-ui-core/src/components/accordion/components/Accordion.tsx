import type { AccordionPropsTw } from '@types';

import { Box } from '@components/box';
import { PropsWithChildren } from 'react';
import { classNameHandler, cn, styleHandler } from '@utils';

import styles from '../styles/Accordion.styles';
import { AccordionProvider } from './AccordionContext';
import { accordionDefaultProps } from '../utils/defaultProps';
import { accordionClassName } from '../styles/Accordion.variants';
import { classNamePropsHandler } from '../../../utils/hybrid/classNamePropsHandler';

/**
 * Accordion component provides a collapsible content area with support for single or multiple items.
 * Automatically detects NativeWind availability and falls back to StyleSheet if needed.
 *
 * @example
 * ```tsx
 * <Accordion type="single" collapsible>
 *   <AccordionItem value="item-1">
 *     <AccordionTrigger>Is it accessible?</AccordionTrigger>
 *     <AccordionContent>Yes. It adheres to the WAI-ARIA design pattern.</AccordionContent>
 *   </AccordionItem>
 * </Accordion>
 * ```
 *
 * @param children - Child AccordionItem components
 * @param type - Whether single or multiple items can be open at once ('single' | 'multiple')
 * @param className - Custom CSS classes for styling
 * @param collapseIcon - Icon to show when item is expanded (default: 'angle-down-solid')
 * @param collapsible - Whether items can be collapsed when type is 'single' (default: false)
 * @param defaultValue - Default open items (string for single, array for multiple)
 * @param disabled - Disable the entire accordion (default: false)
 * @param expandIcon - Icon to show when item is collapsed (default: 'angle-down-solid')
 * @param hasCollapseIcon - Whether to show collapse/expand icons (default: true)
 * @param iconPosition - Position of the expand/collapse icon (default: 'right')
 * @param onValueChange - Callback fired when accordion state changes
 * @param props - Additional ViewProps for the container
 *
 * @see {@link https://github.com/your-repo/chill-ui/tree/main/src/components/accordion/README.md Documentation}
 */
export default function Accordion(props: PropsWithChildren<AccordionPropsTw>) {
  classNamePropsHandler(props, 'Accordion');
  const {
    children,
    className,
    collapseIcon = accordionDefaultProps.collapseIcon,
    collapsible = accordionDefaultProps.collapsible,
    defaultValue,
    disabled = accordionDefaultProps.disabled,
    expandIcon = accordionDefaultProps.expandIcon,
    hasCollapseIcon = accordionDefaultProps.hasCollapseIcon,
    iconPosition = accordionDefaultProps.iconPosition,
    onValueChange,
    style,
    type = accordionDefaultProps.type,
  } = props;

  return (
    <AccordionProvider
      type={type!}
      collapsible={collapsible}
      defaultValue={defaultValue}
      disabled={disabled}
      hasAnimation={false}
      animationDuration={0}
      hasCollapseIcon={hasCollapseIcon ?? true}
      iconPosition={iconPosition}
      expandIcon={expandIcon}
      collapseIcon={collapseIcon}
      onValueChange={onValueChange}
    >
      <Box
        {...classNameHandler(cn(accordionClassName, className))}
        {...styleHandler({ defaultStyle: styles.accordion, style })}
      >
        {children}
      </Box>
    </AccordionProvider>
  );
}
