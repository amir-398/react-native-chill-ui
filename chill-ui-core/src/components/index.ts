export * from './box';
export * from './toast';
export * from './loadingIndicatorsKit';
export * from './accordion';
export * from './dialog';
export * from './ripplePressable';
export * from './animatedBox';
export * from './icon';
export * from './avatar';
export * from './string';
export * from './inputDropdown';
export * from './AutocompleteDropdown';
export * from './highlightString';
export * from './buttonIcon';
export * from './checkbox';
export * from './chip';
export * from './maskedInput';
export * from './dropdownMenu';
export * from './inputSelectDropdown';
export * from './phoneNumberInput';
export * from './placesInput';
export * from './segmentedControl';
export * from './separator';
export * from './skeletons';
export * from './timePicker';
export * from './toggle';
export * from './wrapper';
