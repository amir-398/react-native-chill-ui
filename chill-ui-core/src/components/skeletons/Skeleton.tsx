import { cn } from '@utils';
import { View } from 'react-native';
import { tv } from 'tailwind-variants';

import { SkeletonProps } from '../../types/skeleton.types';

/**
 * Skeleton component that provides loading placeholders with animated pulse effect.
 * Supports multiple variants and sizes for different content types.
 *
 * @example
 * ```tsx
 * // Basic rectangle skeleton
 * <Skeleton variant="rectangle" size="md" />
 *
 * // Circle skeleton for avatar
 * <Skeleton variant="circle" size="lg" />
 *
 * // Text skeleton with custom styling
 * <Skeleton variant="text" size="sm" className="mb-2" />
 * ```
 *
 * @param children - Child components to render inside the skeleton
 * @param className - Custom CSS classes for additional styling
 * @param size - Size variant for the skeleton ('xs' | 'sm' | 'md' | 'lg' | 'xl')
 * @param variant - Shape variant for the skeleton ('rectangle' | 'square' | 'circle' | 'text')
 * @returns Skeleton component with animated pulse effect
 */
export default function Skeleton({ children, className, size = 'md', variant = 'rectangle' }: SkeletonProps) {
  /**
   * Tailwind variants for skeleton styling with size and variant combinations
   * Provides different dimensions and shapes based on props
   */
  const skeletonStyleVariants = tv({
    base: 'bg-gray-300 animate-pulse',
    compoundVariants: [
      // Rectangle + taille spécifique
      { class: 'h-16', size: 'xs', variant: 'rectangle' },
      { class: 'h-24', size: 'sm', variant: 'rectangle' },
      { class: 'h-32', size: 'md', variant: 'rectangle' },
      { class: 'h-48', size: 'lg', variant: 'rectangle' },
      { class: 'h-64', size: 'xl', variant: 'rectangle' },

      // Square + taille spécifique
      { class: 'w-16 h-16', size: 'xs', variant: 'square' },
      { class: 'w-24 h-24', size: 'sm', variant: 'square' },
      { class: 'w-32 h-32', size: 'md', variant: 'square' },
      { class: 'w-48 h-48', size: 'lg', variant: 'square' },
      { class: 'w-64 h-64', size: 'xl', variant: 'square' },

      // Circle + taille spécifique
      { class: 'w-16 h-16', size: 'xs', variant: 'circle' },
      { class: 'w-24 h-24', size: 'sm', variant: 'circle' },
      { class: 'w-32 h-32', size: 'md', variant: 'circle' },
      { class: 'w-48 h-48', size: 'lg', variant: 'circle' },
      { class: 'w-64 h-64', size: 'xl', variant: 'circle' },

      { class: 'h-4', size: 'xs', variant: 'text' },
      { class: 'h-5', size: 'sm', variant: 'text' },
      { class: 'h-6', size: 'md', variant: 'text' },
      { class: 'h-7', size: 'lg', variant: 'text' },
      { class: 'h-8', size: 'xl', variant: 'text' },
    ],
    variants: {
      size: {
        lg: 'h-48',
        md: 'h-32',
        sm: 'h-24',
        xl: 'h-64',
        xs: 'h-16', // Taille par défaut pour les variantes
      },
      variant: {
        circle: 'size-16 rounded-full',
        rectangle: 'w-full rounded-lg',
        square: 'size-16 rounded-lg',
        text: 'w-full rounded-sm',
      },
    },
  });
  return <View className={cn(skeletonStyleVariants({ size, variant }), className)}>{children}</View>;
}
