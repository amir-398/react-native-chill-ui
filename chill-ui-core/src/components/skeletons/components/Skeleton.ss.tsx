import { Animated } from 'react-native';
import { SkeletonPropsSs } from '@types';
import { AnimatedBoxSs } from '@components/animatedBox';
import { useEffect, useRef, PropsWithChildren } from 'react';

import { skeletonSv } from '../styles/Skeleton.ss.styles';

/**
 * Skeleton component that provides loading placeholders with animated pulse effect.
 * Supports multiple variants and sizes for different content types.
 *
 * @example
 * ```tsx
 * // Basic rectangle skeleton
 * <Skeleton variant="rectangle" size="md" />
 * ```
 *
 * @param children - Child components to render inside the skeleton
 * @param size - Size variant for the skeleton ('xs' | 'sm' | 'md' | 'lg' | 'xl')
 * @param style - Style object for additional styling
 * @param variant - Shape variant for the skeleton ('rectangle' | 'square' | 'circle' | 'text')
 * @returns Skeleton component with animated pulse effect
 */
export function Skeleton(props: PropsWithChildren<SkeletonPropsSs>) {
  const { children, size, style, variant, ...rest } = props;
  const pulseAnim = useRef(new Animated.Value(1)).current;

  useEffect(() => {
    const pulse = () => {
      Animated.sequence([
        Animated.timing(pulseAnim, {
          duration: 1000,
          toValue: 0.5,
          useNativeDriver: true,
        }),
        Animated.timing(pulseAnim, {
          duration: 1000,
          toValue: 1,
          useNativeDriver: true,
        }),
      ]).start(() => pulse());
    };

    pulse();
  }, [pulseAnim]);

  return (
    <AnimatedBoxSs style={[skeletonSv({ size, variant }), { opacity: pulseAnim }, style]} {...rest}>
      {children}
    </AnimatedBoxSs>
  );
}

Skeleton.displayName = 'Skeleton';
