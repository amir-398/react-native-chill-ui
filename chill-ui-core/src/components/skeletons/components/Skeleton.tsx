import { Box } from '@components/box';
import { Animated } from 'react-native';
import { SkeletonPropsTw } from '@types';
import { AnimatedBox } from '@components/animatedBox';
import { useEffect, useRef, PropsWithChildren } from 'react';
import { isNativeWindInstalled, classNameHandler, classNamePropsHandler, cn, styleHandler } from '@utils';

import { skeletonTv } from '../styles/Skeleton.tw.styles';
import { skeletonSv } from '../styles/Skeleton.ss.styles';

/**
 * The `<Skeleton />` component provides loading placeholders with animated pulse effect.
 * Supports multiple variants and sizes for different content types.
 *
 *
 * <!-- STORYBOOK_IMPORT_START
 * ```tsx
 * import { Skeleton } from 'react-native-chill-ui';
 * ```
 * STORYBOOK_IMPORT_END -->
 *
 * @example
 * ```tsx
 * <Skeleton variant="rectangle" size="md" />
 * ```
 *
 * @param children - Child components to render inside the skeleton
 * @param className - Custom CSS classes for styling (NativeWind)
 * @param size - Size variant for the skeleton: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
 * @param style - Style object for additional styling
 * @param variant - Shape variant for the skeleton: 'rectangle' | 'square' | 'circle' | 'text'
 */
export function Skeleton(props: PropsWithChildren<SkeletonPropsTw>) {
  classNamePropsHandler(props, 'Skeleton');
  const hasNativeWind = isNativeWindInstalled();
  const { children, className, size, style, variant, ...rest } = props;
  const pulseAnim = useRef(new Animated.Value(1)).current;

  useEffect(() => {
    const pulse = () => {
      if (hasNativeWind) return;
      Animated.sequence([
        Animated.timing(pulseAnim, {
          duration: 1000,
          toValue: 0.5,
          useNativeDriver: true,
        }),
        Animated.timing(pulseAnim, {
          duration: 1000,
          toValue: 1,
          useNativeDriver: true,
        }),
      ]).start(() => pulse());
    };

    pulse();
  }, [pulseAnim, hasNativeWind]);

  const Component = hasNativeWind ? AnimatedBox : Box;

  return (
    <Component
      {...classNameHandler(cn(skeletonTv({ size, variant }), className))}
      {...styleHandler({ defaultStyle: [skeletonSv({ size, variant }), { opacity: pulseAnim }], style })}
      {...rest}
    >
      {children}
    </Component>
  );
}

Skeleton.displayName = 'Skeleton';
