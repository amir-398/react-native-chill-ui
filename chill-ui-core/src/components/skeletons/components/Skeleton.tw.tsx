import { cn } from '@utils';
import { BoxTw } from '@components/box';
import { SkeletonPropsTw } from '@types';
import { PropsWithChildren } from 'react';

import { skeletonTv } from '../styles/Skeleton.tw.styles';

/**
 * Skeleton component that provides loading placeholders with animated pulse effect.
 * Supports multiple variants and sizes for different content types.
 *
 * @example
 * ```tsx
 * // Basic rectangle skeleton
 * <Skeleton variant="rectangle" size="md" />
 * ```
 *
 * @param children - Child components to render inside the skeleton
 * @param className - Custom CSS classes for additional styling
 * @param size - Size variant for the skeleton ('xs' | 'sm' | 'md' | 'lg' | 'xl')
 * @param variant - Shape variant for the skeleton ('rectangle' | 'square' | 'circle' | 'text')
 * @returns Skeleton component with animated pulse effect
 */
export function Skeleton(props: PropsWithChildren<SkeletonPropsTw>) {
  const { children, className, size, variant, ...rest } = props;

  return (
    <BoxTw className={cn(skeletonTv({ size, variant }), className)} {...rest}>
      {children}
    </BoxTw>
  );
}

Skeleton.displayName = 'Skeleton';
