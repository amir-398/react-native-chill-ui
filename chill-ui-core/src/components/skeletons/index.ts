export { Skeleton } from './components/Skeleton';
export { Skeleton as SkeletonTw } from './components/Skeleton.tw';
export { Skeleton as SkeletonSs } from './components/Skeleton.ss';
