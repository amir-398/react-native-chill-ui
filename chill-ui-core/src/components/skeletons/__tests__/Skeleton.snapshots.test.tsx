import React from 'react';
import { render } from '@testing-library/react-native';

import { Skeleton } from '../components/Skeleton';

// Mock Animated
jest.mock('react-native', () => {
  const RN = jest.requireActual('react-native');
  RN.Animated.timing = () => ({
    start: () => {},
  });
  return RN;
});

describe('Skeleton Snapshots', () => {
  it('should match snapshot with basic props', () => {
    const tree = render(<Skeleton />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with rectangle variant', () => {
    const tree = render(<Skeleton variant="rectangle" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with circle variant', () => {
    const tree = render(<Skeleton variant="circle" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with square variant', () => {
    const tree = render(<Skeleton variant="square" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with text variant', () => {
    const tree = render(<Skeleton variant="text" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with xs size', () => {
    const tree = render(<Skeleton size="xs" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with sm size', () => {
    const tree = render(<Skeleton size="sm" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with md size', () => {
    const tree = render(<Skeleton size="md" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with lg size', () => {
    const tree = render(<Skeleton size="lg" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with xl size', () => {
    const tree = render(<Skeleton size="xl" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom className', () => {
    const tree = render(<Skeleton className="custom-class" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with custom style', () => {
    const tree = render(<Skeleton style={{ marginBottom: 8 }} />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with accessibility props', () => {
    const tree = render(<Skeleton accessible accessibilityLabel="Loading skeleton" testID="skeleton" />);
    expect(tree).toMatchSnapshot();
  });

  it('should match snapshot with all props combined', () => {
    const tree = render(
      <Skeleton
        variant="circle"
        size="lg"
        className="custom-class"
        style={{ marginBottom: 16 }}
        testID="skeleton"
        accessible
        accessibilityLabel="Custom skeleton"
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
