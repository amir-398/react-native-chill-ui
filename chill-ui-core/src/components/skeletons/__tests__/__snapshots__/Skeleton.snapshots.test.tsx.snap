// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton Snapshots should match snapshot with accessibility props 1`] = `
<View
  accessibilityLabel="Loading skeleton"
  accessible={true}
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-32"
  collapsable={false}
  testID="skeleton"
/>
`;

exports[`Skeleton Snapshots should match snapshot with all props combined 1`] = `
<View
  accessibilityLabel="Custom skeleton"
  accessible={true}
  className="overflow-hidden bg-gray-300 animate-pulse size-16 rounded-full w-48 h-48 custom-class"
  collapsable={false}
  style={
    {
      "marginBottom": 16,
    }
  }
  testID="skeleton"
/>
`;

exports[`Skeleton Snapshots should match snapshot with basic props 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-32"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with circle variant 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse size-16 rounded-full w-32 h-32"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with custom className 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-32 custom-class"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with custom style 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-32"
  collapsable={false}
  style={
    {
      "marginBottom": 8,
    }
  }
/>
`;

exports[`Skeleton Snapshots should match snapshot with lg size 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-48"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with md size 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-32"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with rectangle variant 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-32"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with sm size 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-24"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with square variant 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse size-16 rounded-lg w-32 h-32"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with text variant 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-sm h-6"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with xl size 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-64"
  collapsable={false}
/>
`;

exports[`Skeleton Snapshots should match snapshot with xs size 1`] = `
<View
  className="overflow-hidden bg-gray-300 animate-pulse w-full rounded-lg h-16"
  collapsable={false}
/>
`;
